<h2>Table of Contents</h2>
<div class="alert alert-block alert-info" style="margin-top: 20px">
    <ul>
    <li><a href="https://#Section_1">Instructions</a></li>
    <li><a href="https://#Section_2">About the Data</a></li>
    <li><a href="https://#Section_3">Importing Data </a></li>
    <li><a href="https://#Section_4">Data Preprocessing</a> </li>
    <li><a href="https://#Section_5">One Hot Encoding </a></li>
    <li><a href="https://#Section_6">Train and Test Data Split </a></li>
    <li><a href="https://#Section_7">Train Logistic Regression, KNN, Decision Tree, SVM, and Linear Regression models and return their appropriate accuracy scores</a></li>
</a></li>

<hr>


# Instructions


In this notebook, you will  practice all the classification algorithms that we have learned in this course.


Below, is where we are going to use the classification algorithms to create a model based on our training data and evaluate our testing data using evaluation metrics learned in the course.

We will use some of the algorithms taught in the course, specifically:

1. Linear Regression
2. KNN
3. Decision Trees
4. Logistic Regression
5. SVM

We will evaluate our models using:

1.  Accuracy Score
2.  Jaccard Index
3.  F1-Score
4.  LogLoss
5.  Mean Absolute Error
6.  Mean Squared Error
7.  R2-Score

Finally, you will use your models to generate the report at the end. 


# About The Dataset


The original source of the data is Australian Government's Bureau of Meteorology and the latest data can be gathered from [http://www.bom.gov.au/climate/dwo/](http://www.bom.gov.au/climate/dwo/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkML0101ENSkillsNetwork20718538-2022-01-01).

The dataset to be used has extra columns like 'RainToday' and our target is 'RainTomorrow', which was gathered from the Rattle at [https://bitbucket.org/kayontoga/rattle/src/master/data/weatherAUS.RData](https://bitbucket.org/kayontoga/rattle/src/master/data/weatherAUS.RData?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkML0101ENSkillsNetwork20718538-2022-01-01)




This dataset contains observations of weather metrics for each day from 2008 to 2017. The **weatherAUS.csv** dataset includes the following fields:

| Field         | Description                                           | Unit            | Type   |
| ------------- | ----------------------------------------------------- | --------------- | ------ |
| Date          | Date of the Observation in YYYY-MM-DD                 | Date            | object |
| Location      | Location of the Observation                           | Location        | object |
| MinTemp       | Minimum temperature                                   | Celsius         | float  |
| MaxTemp       | Maximum temperature                                   | Celsius         | float  |
| Rainfall      | Amount of rainfall                                    | Millimeters     | float  |
| Evaporation   | Amount of evaporation                                 | Millimeters     | float  |
| Sunshine      | Amount of bright sunshine                             | hours           | float  |
| WindGustDir   | Direction of the strongest gust                       | Compass Points  | object |
| WindGustSpeed | Speed of the strongest gust                           | Kilometers/Hour | object |
| WindDir9am    | Wind direction averaged of 10 minutes prior to 9am    | Compass Points  | object |
| WindDir3pm    | Wind direction averaged of 10 minutes prior to 3pm    | Compass Points  | object |
| WindSpeed9am  | Wind speed averaged of 10 minutes prior to 9am        | Kilometers/Hour | float  |
| WindSpeed3pm  | Wind speed averaged of 10 minutes prior to 3pm        | Kilometers/Hour | float  |
| Humidity9am   | Humidity at 9am                                       | Percent         | float  |
| Humidity3pm   | Humidity at 3pm                                       | Percent         | float  |
| Pressure9am   | Atmospheric pressure reduced to mean sea level at 9am | Hectopascal     | float  |
| Pressure3pm   | Atmospheric pressure reduced to mean sea level at 3pm | Hectopascal     | float  |
| Cloud9am      | Fraction of the sky obscured by cloud at 9am          | Eights          | float  |
| Cloud3pm      | Fraction of the sky obscured by cloud at 3pm          | Eights          | float  |
| Temp9am       | Temperature at 9am                                    | Celsius         | float  |
| Temp3pm       | Temperature at 3pm                                    | Celsius         | float  |
| RainToday     | If there was rain today                               | Yes/No          | object |
| RainTomorrow  | If there is rain tomorrow                             | Yes/No          | float  |

Column definitions were gathered from [http://www.bom.gov.au/climate/dwo/IDCJDW0000.shtml](http://www.bom.gov.au/climate/dwo/IDCJDW0000.shtml?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkML0101ENSkillsNetwork20718538-2022-01-01)



## **Import the required libraries**


In [2]:
# Surpress warnings:
def warn(*args, **kwargs):
    pass
import warnings
warnings.warn = warn
import warnings
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import sklearn.metrics as metrics
import time
import gc, sys
%matplotlib inline
warnings.filterwarnings('ignore')
from sklearn.metrics import mean_absolute_error
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score
from sklearn.neighbors import KNeighborsRegressor
from sklearn.linear_model import LogisticRegression
from sklearn import preprocessing
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn import svm
from sklearn.preprocessing import normalize, StandardScaler, MinMaxScaler
from sklearn.utils.class_weight import compute_sample_weight
from sklearn.metrics import mean_squared_error
from sklearn.metrics import jaccard_score
from sklearn.metrics import f1_score
from sklearn.metrics import log_loss
from sklearn.metrics import confusion_matrix, accuracy_score
from __future__ import print_function


### Importing the Dataset


In [18]:
path='https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-ML0101EN-SkillUp/labs/ML-FinalAssignment/Weather_Data.csv'
#await download(path, "Weather_Data.csv")
#filename ="Weather_Data.csv"

import requests

def download(url, filename):
    # Make a GET request to the specified URL
    response = requests.get(url)
    
    # Check if the request was successful (status code 200)
    if response.status_code == 200:
        # Open the file specified by 'filename' in binary write mode
        with open(filename, "wb") as f:
            # Write the content of the response to the file
            f.write(response.content)
            # Print a message indicating that the download is completed
            print("Download completed.")
    else:
        # Print an error message if the request fails
        print(f"Failed to download from {url}. Status code: {response.status_code}")

download("https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-ML0101EN-SkillUp/labs/ML-FinalAssignment/Weather_Data.csv", "Weather_Data.csv")

df = pd.read_csv("Weather_Data.csv")
df.head()

Download completed.


Date  MinTemp  MaxTemp  Rainfall  Evaporation  Sunshine WindGustDir  \
0  2/1/2008     19.5     22.4      15.6          6.2       0.0           W   
1  2/2/2008     19.5     25.6       6.0          3.4       2.7           W   
2  2/3/2008     21.6     24.5       6.6          2.4       0.1           W   
3  2/4/2008     20.2     22.8      18.8          2.2       0.0           W   
4  2/5/2008     19.7     25.7      77.4          4.8       0.0           W   

   WindGustSpeed WindDir9am WindDir3pm  ...  Humidity9am  Humidity3pm  \
0             41          S        SSW  ...           92           84   
1             41          W          E  ...           83           73   
2             41        ESE        ESE  ...           88           86   
3             41        NNE          E  ...           83           90   
4             41        NNE          W  ...           88           74   

   Pressure9am  Pressure3pm  Cloud9am  Cloud3pm  Temp9am  Temp3pm  RainToday  \
0       1017.6       1017.4         8         8     20.7     20.9        Yes   
1       1017.9       1016.4         7         7     22.4     24.8        Yes   
2       1016.7       1015.6         7         8     23.5     23.0        Yes   
3       1014.2       1011.8         8         8     21.4     20.9        Yes   
4       1008.3       1004.8         8         8     22.5     25.5        Yes   

   RainTomorrow  
0           Yes  
1           Yes  
2           Yes  
3           Yes  
4           Yes  

[5 rows x 22 columns]

### Data Preprocessing


In [23]:
# First, we need to perform one hot encoding to convert categorical variables to binary variables.
df_sydney_processed = pd.get_dummies(data=df, columns=['RainToday', 'WindGustDir', 'WindDir9am', 'WindDir3pm'])

Next, we replace the values of the 'RainTomorrow' column changing them from a categorical column to a binary column. We do not use the `get_dummies` method because we would end up with two columns for 'RainTomorrow' and we do not want, since 'RainTomorrow' is our target.


In [24]:
df_sydney_processed.replace(['No', 'Yes'], [0,1], inplace=True)

### Training Data and Test Data


Now, we set our 'features' or x values and our Y or target variable.


In [25]:
df_sydney_processed.drop('Date',axis=1,inplace=True)

df_sydney_processed = df_sydney_processed.astype(float)


features = df_sydney_processed.drop(columns='RainTomorrow', axis=1)
Y = df_sydney_processed['RainTomorrow']

### Linear Regression


In [26]:
# Split the data into training and testing sets
X_train, X_test, Y_train, Y_test = train_test_split(features, Y, test_size=0.2, random_state=10)

# Print the shapes of the training and testing sets to verify the split
print("Training set - Features:", X_train.shape, "Training set - Target:", Y_train.shape)
print("Testing set - Features:", X_test.shape, "Testing set - Target:", Y_test.shape)

# Create and fit the linear regression model
model = LinearRegression()
model.fit(X_train, Y_train)

# Predict on training data
Y_train_pred = model.predict(X_train)

# Predict on test data
Y_test_pred = model.predict(X_test)

# Calculate R² for the training data
r2_train = r2_score(Y_train, Y_train_pred)
print(f"R² for the training data: {r2_train}")

# Calculate R² for the test data
r2_test = r2_score(Y_test, Y_test_pred)
print(f"R² for the test data: {r2_test}")

Training set - Features: (2616, 66) Training set - Target: (2616,)
Testing set - Features: (655, 66) Testing set - Target: (655,)
R² for the training data: 0.38691299347235053
R² for the test data: 0.4271321073623009


Interpretation of R² Values
The R² (coefficient of determination) values you have obtained are as follows:

R² for the training data: 0.3869
R² for the test data: 0.4271
Explanation of Results
R² Value Interpretation:

R² for the training data (0.3869): This value indicates that approximately 38.69% of the variance in the target variable (price) is explained by the features in the training dataset. While this indicates a moderate level of explanatory power, it suggests that a significant portion of the variance is not captured by the model.
R² for the test data (0.4271): This value indicates that approximately 42.71% of the variance in the target variable is explained by the features in the test dataset. This is slightly higher than the training R², suggesting that the model performs marginally better on unseen data.
Overfitting and Underfitting:

Underfitting: Underfitting occurs when a model is too simple and cannot capture the underlying patterns in the data. This typically results in low R² values for both the training and test datasets. In your case, the moderate R² values (around 0.39 and 0.43) suggest that the model might be underfitting slightly, as it is not explaining a large portion of the variance in the data. This indicates that the model may be too simple or that the features selected do not capture all the relevant information.
Overfitting: Overfitting occurs when a model is too complex and captures not only the underlying patterns but also the noise in the training data. This often results in a high R² value for the training data and a significantly lower R² value for the test data. In your case, the R² values for both the training and test datasets are relatively close, which suggests that the model is not overfitting. The fact that the test R² is slightly higher than the training R² could be due to variability in the data or the selection of the test set.
Possible Actions:
Feature Engineering: Consider adding more relevant features or creating polynomial features to capture more complex relationships in the data.
Model Complexity: Explore more complex models or regularization techniques to improve the model's ability to capture underlying patterns without overfitting.
Data Quality: Ensure that the data is clean, and consider whether additional data could help improve the model's performance.
In summary, while the model does not seem to suffer from overfitting, the moderate R² values indicate that there is room for improvement in capturing the variance in the target variable. Exploring additional features, more complex models, or different modeling techniques could help achieve better performance.

In [32]:
# Split the data into training and testing sets
X_train, X_test, Y_train, Y_test = train_test_split(features, Y, test_size=0.2, random_state=10)

# Create a linear regression model
model = LinearRegression()

# Train the model on the training data
model.fit(X_train, Y_train)

# Make predictions on the testing data
Y_pred = model.predict(X_test)

# Evaluate the model
mse = mean_squared_error(Y_test, Y_pred)
mae = mean_absolute_error(Y_test, Y_pred)
r2 = r2_score(Y_test, Y_pred)


# Use the trained model to make predictions on the testing data
predictions = model.predict(X_test)

# Show the MAE, MSE, and R2 in a tabular format using data frame for the linear model.

# Create a DataFrame to store the evaluation metrics
evaluation_df = pd.DataFrame({
    'Metric': ['Mean Absolute Error (MAE)', 'Mean Squared Error (MSE)', 'R-squared (R2)'],
    'Value': [mae, mse, r2]
})

# Print the DataFrame
print("Linear Regression Model Evaluation:")
print(evaluation_df)

Linear Regression Model Evaluation:
                      Metric     Value
0  Mean Absolute Error (MAE)  0.256318
1   Mean Squared Error (MSE)  0.115721
2             R-squared (R2)  0.427132


### KNN


In [29]:
from sklearn.metrics import accuracy_score, jaccard_score, f1_score

# Split the data into training and testing sets
X_train, X_test, Y_train, Y_test = train_test_split(features, Y, test_size=0.2, random_state=10)

# Create a KNN regression model with n_neighbors=4
knn_model = KNeighborsRegressor(n_neighbors=4)

# Train the KNN model on the training data
knn_model.fit(X_train, Y_train)

# Make predictions on the training data
Y_train_pred = knn_model.predict(X_train)

# Calculate evaluation metrics for the training data
train_mae = mean_absolute_error(Y_train, Y_train_pred)
train_mse = mean_squared_error(Y_train, Y_train_pred)
train_r2 = r2_score(Y_train, Y_train_pred)

# Print evaluation metrics for the training data
print("Training Data Evaluation:")
print("Mean Absolute Error (MAE):", train_mae)
print("Mean Squared Error (MSE):", train_mse)
print("R-squared (R2):", train_r2)

# Make predictions on the test data
Y_test_pred = knn_model.predict(X_test)

# Calculate evaluation metrics for the test data
test_mae = mean_absolute_error(Y_test, Y_test_pred)
test_mse = mean_squared_error(Y_test, Y_test_pred)
test_r2 = r2_score(Y_test, Y_test_pred)

# Print evaluation metrics for the test data
print("Test Data Evaluation:")
print("Mean Absolute Error (MAE):", test_mae)
print("Mean Squared Error (MSE):", test_mse)
print("R-squared (R2):", test_r2)

Training Data Evaluation:
Mean Absolute Error (MAE): 0.17584097859327216
Mean Squared Error (MSE): 0.08887614678899082
R-squared (R2): 0.531206283263258
Test Data Evaluation:
Mean Absolute Error (MAE): 0.2316793893129771
Mean Squared Error (MSE): 0.1441793893129771
R-squared (R2): 0.28624847110680307


In the previous we got an error of



 ( 1 # Using the `predict` method on the testing data (`x_test`) and save it to the array `predictions`.
      2 # Use the trained KNN model to make predictions on the testing data
----> 3 predictions = KNN.predict(X_test)

File c:\Users\POP\anaconda3\envs\myenv\Lib\site-packages\sklearn\neighbors\_regression.py:242, in KNeighborsRegressor.predict(self, X)
    226 """Predict the target for the provided data.
    227 
    228 Parameters
   (...)
    237     Target values.
    238 """
    239 if self.weights == "uniform":
    240     # In that case, we do not need the distances to perform
    241     # the weighting so we do not compute them.
--> 242     neigh_ind = self.kneighbors(X, return_distance=False)
    243     neigh_dist = None
    244 else:

File c:\Users\POP\anaconda3\envs\myenv\Lib\site-packages\sklearn\neighbors\_base.py:850, in KNeighborsMixin.kneighbors(self, X, n_neighbors, return_distance)
    843 use_pairwise_distances_reductions = (
    844     self._fit_method == "brute"
    845     and ArgKmin.is_usable_for(
...
--> 646 config = get_config().split()
    647 if config[0] == b"OpenBLAS":
    648     return config[1].decode("utf-8")

AttributeError: 'NoneType' object has no attribute 'split')


the interpretation: 
This error seems to be related to the configuration of OpenBLAS, a library often used for numerical computations in Python, particularly with libraries like NumPy and scikit-learn. It appears that there might be an issue with retrieving the configuration, leading to a NoneType error when attempting to split it.


we solve it by (pip install --upgrade --force-reinstall scikit-learn) # successfully upgraded scikit-learn in your myenv environment



### Decision Tree


In [30]:
# Create a decision tree classifier
Tree = DecisionTreeClassifier()

# Train the decision tree classifier on the training data
Tree.fit(X_train, Y_train)

# Use the predict method on the testing data and save it to the array `predictions`
predictions = Tree.predict(X_test)

# Calculate the value for each metric using the appropriate function
Tree_Accuracy_Score = accuracy_score(Y_test, predictions)
Tree_JaccardIndex = jaccard_score(Y_test, predictions)
Tree_F1_Score = f1_score(Y_test, predictions)

# Print the evaluation metrics
print("Decision Tree Classifier Model Evaluation:")
print("Accuracy Score:", Tree_Accuracy_Score)
print("Jaccard Index:", Tree_JaccardIndex)
print("F1 Score:", Tree_F1_Score)

Decision Tree Classifier Model Evaluation:
Accuracy Score: 0.7648854961832061
Jaccard Index: 0.4166666666666667
F1 Score: 0.5882352941176471


### Logistic Regression


In [15]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, jaccard_score, f1_score, log_loss

# Split the data into training and testing sets with test_size=0.2 and random_state=1
x_train, x_test, y_train, y_test = train_test_split(features, Y, test_size=0.2, random_state=1)

# Create and train a LogisticRegression model with solver set to 'liblinear'
LR = LogisticRegression(solver='liblinear')
LR.fit(x_train, y_train)

# Use the trained model to make predictions on the testing data
predictions = LR.predict(x_test)
predict_proba = LR.predict_proba(x_test)

# Calculate the evaluation metrics
LR_Accuracy_Score = accuracy_score(y_test, predictions)
LR_JaccardIndex = jaccard_score(y_test, predictions)
LR_F1_Score = f1_score(y_test, predictions)
LR_Log_Loss = log_loss(y_test, predict_proba)

# Print the evaluation metrics
print("Logistic Regression Model Evaluation:")
print(f"Accuracy Score: {LR_Accuracy_Score}")
print(f"Jaccard Index: {LR_JaccardIndex}")
print(f"F1 Score: {LR_F1_Score}")
print(f"Log Loss: {LR_Log_Loss}")


Logistic Regression Model Evaluation:
Accuracy Score: 0.8351145038167939
Jaccard Index: 0.5045871559633027
F1 Score: 0.6707317073170732
Log Loss: 0.3815916633912949


### SVM


In [33]:
from sklearn.svm import SVC

# Create an SVM model (Support Vector Machine)
SVM = SVC() 

# Train the SVM model on the training data
SVM.fit(X_train, Y_train)

# Make predictions on the testing data
predictions = SVM.predict(X_test)

# Evaluate the SVM model
SVM_Accuracy_Score = accuracy_score(Y_test, predictions)
SVM_JaccardIndex = jaccard_score(Y_test, predictions)
SVM_F1_Score = f1_score(Y_test, predictions)

# Print evaluation metrics for SVM
print("SVM Model Evaluation:")
print("Accuracy Score:", SVM_Accuracy_Score)
print("Jaccard Index:", SVM_JaccardIndex)
print("F1 Score:", SVM_F1_Score)

SVM Model Evaluation:
Accuracy Score: 0.7190839694656489
Jaccard Index: 0.0
F1 Score: 0.0


### Report


In [39]:
print("Report")

# Create a DataFrame to compare the evaluation metrics
comparison_df = pd.DataFrame({
    'Model': ['Linear Regression', 'KNN Regression', 'Logistic Regression', 'Decision Tree Classifier', 'Support Vector Machine'],
    'MSE': [mse, test_mse, None, None, None],
    'MAE': [mae, test_mae, None, None, None],
    'R2 Score': [r2, test_r2, None, None, None],
    'Accuracy': [None, None, LR_Accuracy_Score, Tree_Accuracy_Score, SVM_Accuracy_Score],
    'Jaccard Index': [None, None, LR_JaccardIndex, Tree_JaccardIndex, SVM_JaccardIndex],
    'F1 Score': [None, None, LR_F1_Score, Tree_F1_Score, SVM_F1_Score],
    'Log Loss': [None, None, LR_Log_Loss, None, None],
})

# Print the comparison DataFrame
print("\nModel Comparison:")
print(comparison_df)


Report

Model Comparison:
                      Model       MSE       MAE  R2 Score  Accuracy  \
0         Linear Regression  0.115721  0.256318  0.427132       NaN   
1            KNN Regression  0.144179  0.231679  0.286248       NaN   
2       Logistic Regression       NaN       NaN       NaN  0.835115   
3  Decision Tree Classifier       NaN       NaN       NaN  0.764885   
4    Support Vector Machine       NaN       NaN       NaN  0.719084   

   Jaccard Index  F1 Score  Log Loss  
0            NaN       NaN       NaN  
1            NaN       NaN       NaN  
2       0.504587  0.670732  0.381592  
3       0.416667  0.588235       NaN  
4       0.000000  0.000000       NaN  


### Conclusion: the highest accuracy related to the Logistic Regression model.
### so if are going to predect the Rainfall in spacific area you should use the Logistic Regression model.

### Linear Regression
Mean Squared Error (MSE): 0.115721
Mean Absolute Error (MAE): 0.256318
R² Score: 0.427132
Interpretation:

The Linear Regression model has a moderate R² score, indicating that approximately 42.7% of the variability in the target variable is explained by the model.
The errors (MSE and MAE) are relatively low, suggesting the model's predictions are reasonably close to the actual values.

### KNN Regression
Mean Squared Error (MSE): 0.144179
Mean Absolute Error (MAE): 0.231679
R² Score: 0.286248
Interpretation:

The KNN Regression model has a lower R² score (28.6%), indicating it explains less variance in the target variable compared to Linear Regression.
The errors are slightly higher than those of Linear Regression, suggesting it might not perform as well on this dataset.

### Logistic Regression
Accuracy: 0.835115
Jaccard Index: 0.504587
F1 Score: 0.670732
Log Loss: 0.381592
Interpretation:

The Logistic Regression model has a high accuracy (83.5%), indicating it correctly classifies the majority of the test instances.
The Jaccard Index (0.504587) and F1 Score (0.670732) show that the model performs well in terms of precision and recall balance.
The Log Loss is relatively low, suggesting good performance in terms of probability predictions.

### Decision Tree Classifier
Accuracy: 0.764885
Jaccard Index: 0.416667
F1 Score: 0.588235
Interpretation:

The Decision Tree Classifier has a decent accuracy (76.5%) but lower than Logistic Regression.
The Jaccard Index and F1 Score are also lower compared to Logistic Regression, indicating it may not be as effective in balancing precision and recall.

### Support Vector Machine (SVM)
Accuracy: 0.719084
Jaccard Index: 0.000000
F1 Score: 0.000000
Interpretation:

The SVM model has an accuracy of 71.9%, which is lower than both Logistic Regression and Decision Tree Classifier.
The Jaccard Index and F1 Score being zero indicate that the SVM model failed to identify any positives, likely predicting all instances as the negative class.

### Overall Comparison
Best Model for Regression: Based on R² Score, Linear Regression performs better than KNN Regression, explaining more variance in the target variable.
Best Model for Classification: Logistic Regression outperforms both Decision Tree Classifier and SVM in terms of accuracy, Jaccard Index, F1 Score, and Log Loss.

### Recommendations
Regression Tasks: Use Linear Regression, as it provides better error metrics and explains more variability in the target.
Classification Tasks: Use Logistic Regression, as it has the highest accuracy and balanced performance metrics.
Note: The SVM model’s poor performance could be due to suboptimal parameter choices or the nature of the dataset, suggesting a need for further tuning or a different kernel function.